### PR TITLE
Batch Polymorphism and Abstract Factory

### PR DESCRIPTION
**Pull request by Akvinikym**
### Description of the Change

This PR introduces two things:
- interface for batch class and thus corresponding changes in all touched files
- abstract factory for creating batches from transactions and corresponding test

### Benefits

Run-time polymorphism for batches.

### Possible Drawbacks 

None.

### Usage Examples or Tests *[optional]*

New test target `transaction_batch_unsafe_factory_test` was added.

### CODE REVIEW

File: shared_model/interfaces/iroha_internal/transaction_batch.hpp [hunk 1]
```diff
@@ -14,57 +14,59 @@
...
-      bool hasAllSignatures() const;
+      virtual bool hasAllSignatures() const = 0;
 
-      bool operator==(const TransactionBatch &rhs) const;
+      /**
+       * Checks of two transaction batches are the same
+       * @param rhs - another batch
+       * @return true if they are equal, false otherwise
+       */
+      virtual bool operator==(const TransactionBatch &rhs) const = 0;
```
Comment (lebdron) [inline shared_model/interfaces/iroha_internal/transaction_batch.hpp]: Inherit from `ModelPrimitive` instead?

File: shared_model/interfaces/iroha_internal/transaction_batch_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#define IROHA_TRANSACTION_BATCH_IMPL_HPP
+
+#include "interfaces/iroha_internal/transaction_batch.hpp"
+
+namespace shared_model {
+  namespace interface {
+
+    class TransactionBatchImpl : public TransactionBatch {
+     public:
+      TransactionBatchImpl() = delete;
```
Comment (lebdron) [inline shared_model/interfaces/iroha_internal/transaction_batch_impl.hpp]: Is it necessary to delete the default constructor if another ctor is already declared? https://github.com/hyperledger/iroha/blob/e7255a34b174b4efb26151a1c7e14cb0eab36125/shared_model/interfaces/iroha_internal/transaction_batch_impl.hpp#L23-L24

File: test/module/shared_model/backend_proto/proto_batch_test.cpp [hunk 3]
```diff
@@ -65,8 +65,12 @@ TEST(TransactionBatchTest, CreateTransactionBatchWhenValid) {
   auto transaction_batch =
       interface::TransactionBatchFactory::createTransactionBatch(
           txs, validation::DefaultUnsignedTransactionsValidator());
-  ASSERT_TRUE(framework::expected::val(transaction_batch))
-      << framework::expected::err(transaction_batch).value().error;
+  transaction_batch.match(
```
Comment (lebdron) [inline test/module/shared_model/backend_proto/proto_batch_test.cpp]: Please fix `val` and `err` to accept non-copyable argument and reuse the methods. Same for changes below.

File: shared_model/interfaces/iroha_internal/transaction_batch_factory.hpp [hunk 4]
```diff
@@ -33,7 +33,8 @@ namespace shared_model {
        */
       template <typename TransactionValidator,
                 typename FieldValidator = validation::FieldValidator>
-      static iroha::expected::Result<TransactionBatch, std::string>
+      static iroha::expected::Result<std::unique_ptr<TransactionBatch>,
```
Comment (lebdron) [inline shared_model/interfaces/iroha_internal/transaction_batch_factory.hpp]: Unsafe factory should be removed, and this factory should be refactored with polymorphism, similar to existing ProposalFactory.

File: shared_model/interfaces/iroha_internal/transaction_batch.hpp [hunk 5]
```diff
@@ -14,57 +14,59 @@
...
-      bool addSignature(size_t number_of_tx,
-                        const shared_model::crypto::Signed &signed_blob,
-                        const shared_model::crypto::PublicKey &public_key);
+      virtual bool addSignature(
+          size_t number_of_tx,
+          const shared_model::crypto::Signed &signed_blob,
+          const shared_model::crypto::PublicKey &public_key) = 0;
 
      private:
       types::SharedTxsCollectionType transactions_;
```
Comment (l4l) [inline shared_model/interfaces/iroha_internal/transaction_batch.hpp]: Most probably shouldn't be there

File: shared_model/interfaces/iroha_internal/transaction_batch_factory.hpp (hunk 4 above)
Comment (l4l) [inline shared_model/interfaces/iroha_internal/transaction_batch_factory.hpp]: What's the profit of class? Maybe just put functions in namespace?

File: test/framework/batch_helper.hpp [hunk 6]
```diff
@@ -197,7 +197,10 @@ namespace framework {
       auto result_batch = shared_model::interface::TransactionBatchFactory::
           createTransactionBatch(txs, TxsValidator());
 
-      return framework::expected::val(result_batch).value().value;
+      return boost::get<iroha::expected::Value<
```
Comment (l4l) [inline test/framework/batch_helper.hpp]: Is it possible to overload expected::val for unique_ptr? Or Result itself require type to be copy-able?

File: test/module/shared_model/factories/transaction_batch_unsafe_factory_test.cpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+   */
+  types::SharedTxsCollectionType createTransactions(int tx_number) {
+    types::SharedTxsCollectionType txs;
+    for (auto i = 0; i < tx_number; ++i) {
+      txs.push_back(createTransaction(i + 1));
+    }
+    return txs;
+  }
+};
+
```
Comment (l4l) [inline test/module/shared_model/factories/transaction_batch_unsafe_factory_test.cpp]: Add also test case for empty transaction list please

File: shared_model/interfaces/iroha_internal/transaction_batch_factory.hpp (hunk 4 above)
Comment (lebdron) [reply shared_model/interfaces/iroha_internal/transaction_batch_factory.hpp]: These methods should be virtual, so they are required to be in a class.

Comment (l4l) [reply shared_model/interfaces/iroha_internal/transaction_batch_factory.hpp]: but they aren't :(
would be nice to add a todo mark then

File: test/module/irohad/torii/torii_service_test.cpp [hunk 8]
```diff
@@ -610,8 +610,9 @@ TEST_F(ToriiServiceTest, FailedListOfTxs) {
         ASSERT_EQ(toriiResponse.tx_status(),
                   iroha::protocol::TxStatus::STATELESS_VALIDATION_FAILED);
         auto msg = toriiResponse.error_message();
-        ASSERT_THAT(toriiResponse.error_message(),
-                    HasSubstr("bad timestamp: sent from future"));
+        ASSERT_THAT(
+            toriiResponse.error_message(),
+            HasSubstr("Transaction should contain at least one signature"));
```
Comment (l4l) [inline test/module/irohad/torii/torii_service_test.cpp]: That a strange fix

File: test/module/shared_model/backend_proto/proto_batch_test.cpp [hunk 9]
```diff
@@ -9,49 +9,102 @@
...
-      clone(framework::batch::prepareTransactionBuilder("invalid#@account",
-                                                        created_time)
-                .build()));
-}
+class TransactionBatchTest : public Test {
+ public:
+  /**
+   * Creates valid unsigned transaction
+   * @param created_time assigned to transactions
+   * @return std::shared_ptr<interface::Transaction> containing valid unsigned
```
Comment (l4l) [inline test/module/shared_model/backend_proto/proto_batch_test.cpp]: Type is redundant since:
1. Function is simple and it's easy to grasp
2. May lead to inconsistency in case function changes

File: test/module/shared_model/backend_proto/proto_batch_test.cpp [hunk 10]
```diff
@@ -9,49 +9,102 @@
...
+    return std::shared_ptr<interface::Transaction>(
+        clone(framework::batch::prepareTransactionBuilder("valid@account",
+                                                          created_time)
+                  .build()));
+  }
+
+  /**
+   * Creates invalid unsigned transaction
+   * @param created_time assigned to transactions
+   * @return std::shared_ptr<interface::Transaction> containing invalid unsigned
```
Comment (l4l) [inline test/module/shared_model/backend_proto/proto_batch_test.cpp]: same here

File: test/framework/batch_helper.hpp [hunk 11]
```diff
@@ -207,23 +207,24 @@ namespace framework {
...
-      return shared_model::interface::TransactionBatchFactory::
-          createTransactionBatch(
-                 tx,
-                 shared_model::validation::DefaultSignedTransactionValidator())
-              .match(
-                  [](const iroha::expected::Value<
-                      shared_model::interface::TransactionBatch> &value) {
-                    return std::make_shared<
-                    shared_model::interface::TransactionBatch>(value.value);
+      std::shared_ptr<shared_model::interface::TransactionBatchFactory>
```
Comment (l4l) [inline test/framework/batch_helper.hpp]: Is type required?

File: test/framework/batch_helper.hpp (hunk 6 above)
Comment (l4l) [reply test/framework/batch_helper.hpp]: I noticed some changes for expected::val/err but this line remains the same. Could you explain this?

File: shared_model/interfaces/iroha_internal/transaction_sequence_factory.hpp [hunk 12]
```diff
@@ -9,17 +9,20 @@
...
   namespace interface {
 
     /**
      * Provides a method that creates a transaction sequence from  a collection
-     * of transactions. Field validator is used by default
+     * of transactions
      */
     class TransactionSequenceFactory {
+     private:
+      static const std::unique_ptr<TransactionBatchFactory> batch_factory_;
```
Comment (l4l) [inline shared_model/interfaces/iroha_internal/transaction_sequence_factory.hpp]: Looks illegal:
1. Private member in the beginning of the definition
2. Static + unique_ptr
3. Not parametrized in ctor?

File: shared_model/interfaces/iroha_internal/transaction_batch_factory_impl.cpp [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+  bool has_at_least_one_signature =
+      std::any_of(transactions.begin(), transactions.end(), [](const auto tx) {
+        return not boost::empty(tx->signatures());
+      });
+  if (not has_at_least_one_signature) {
+    batch_reason.second.emplace_back(
+        "Transaction batch should contain at least one signature");
+  }
+
+  if (not batch_reason.second.empty()) {
```
Comment (l4l) [inline shared_model/interfaces/iroha_internal/transaction_batch_factory_impl.cpp]: It seems that validation that was before (`TransactionBatchFactory::createTransactionBatch`) is gone. Could you explain where is it moved?

File: shared_model/interfaces/iroha_internal/transaction_batch_factory_impl.cpp [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "interfaces/iroha_internal/transaction_batch_factory_impl.hpp"
+
+#include "interfaces/iroha_internal/transaction_batch_impl.hpp"
+#include "interfaces/transaction.hpp"
+#include "validators/answer.hpp"
+
+shared_model::interface::TransactionBatchFactoryImpl::FactoryImplResult
```
Comment (l4l) [inline shared_model/interfaces/iroha_internal/transaction_batch_factory_impl.cpp]: Wrap with namespaces plz

File: irohad/ordering/impl/single_peer_ordering_service.cpp [hunk 15]
```diff
@@ -74,8 +75,11 @@ namespace iroha {
           batch_prop_mutex_);
 
       current_size_.fetch_add(batch.transactions().size());
-      queue_.push(std::make_unique<shared_model::interface::TransactionBatch>(
-          std::move(batch)));
+      queue_.push(
+          std::make_unique<shared_model::interface::TransactionBatchImpl>(
+              std::move(
```
Comment (l4l) [inline irohad/ordering/impl/single_peer_ordering_service.cpp]: Is that `std::move` required? I guess rvalue-ref type is enough in cast, doesn't it?

File: shared_model/interfaces/iroha_internal/transaction_batch_impl.cpp [hunk 16]
```diff
@@ -74,5 +74,16 @@ namespace shared_model {
                          });
     }
 
+    TransactionBatch *TransactionBatchImpl::clone() const {
+      const auto &original_txs = this->transactions_;
+      types::SharedTxsCollectionType copy_txs;
+      std::for_each(std::begin(original_txs),
```
Comment (l4l) [inline shared_model/interfaces/iroha_internal/transaction_batch_impl.cpp]: It's better use std::accumulate, I believe

File: shared_model/interfaces/iroha_internal/transaction_sequence_factory.cpp [hunk 17]
```diff
@@ -28,28 +36,51 @@ namespace shared_model {
...
         result.addReason(std::make_pair(
             "Transaction collection error",
             std::vector<std::string>{"sequence can not be empty"}));
       }
       for (const auto &tx : transactions) {
+        // perform stateless validation checks
+        validation::ReasonsGroupType reason;
+        reason.first = "Transaction: ";
+        // check signatures and their validness
+        if (not boost::empty(tx->signatures())) {
```
Comment (l4l) [inline shared_model/interfaces/iroha_internal/transaction_sequence_factory.cpp]: Revert conditional please, so that there's no `else` and code become more linear

File: irohad/ordering/impl/ordering_service_transport_grpc.hpp [hunk 18]
```diff
@@ -53,6 +54,8 @@ namespace iroha {
       std::weak_ptr<iroha::network::OrderingServiceNotification> subscriber_;
       std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>>
           async_call_;
+      std::unique_ptr<shared_model::interface::TransactionBatchFactory>
```
Comment (lebdron) [inline irohad/ordering/impl/ordering_service_transport_grpc.hpp]: `shared_ptr` might work better since this factory will be reused in several components.

File: irohad/ordering/impl/ordering_service_transport_grpc.cpp [hunk 19]
```diff
@@ -84,4 +80,6 @@ void OrderingServiceTransportGrpc::publishProposal(
 OrderingServiceTransportGrpc::OrderingServiceTransportGrpc(
     std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>>
         async_call)
-    : async_call_(std::move(async_call)) {}
+    : async_call_(std::move(async_call)),
+      batch_factory_(std::make_unique<
+                     shared_model::interface::TransactionBatchFactoryImpl>()) {}
```
Comment (lebdron) [inline irohad/ordering/impl/ordering_service_transport_grpc.cpp]: Please pass it from `application.cpp` instead, so that impl is only included in `application.cpp`.

File: irohad/ordering/impl/single_peer_ordering_service.cpp [hunk 20]
```diff
@@ -74,8 +75,11 @@ namespace iroha {
           batch_prop_mutex_);
 
       current_size_.fetch_add(batch.transactions().size());
-      queue_.push(std::make_unique<shared_model::interface::TransactionBatch>(
-          std::move(batch)));
+      queue_.push(
+          std::make_unique<shared_model::interface::TransactionBatchImpl>(
```
Comment (lebdron) [inline irohad/ordering/impl/single_peer_ordering_service.cpp]: It is better to change `shared_model::interface::TransactionBatch &&` to `std::make_unique<shared_model::interface::TransactionBatch>` in method interface.

File: shared_model/interfaces/iroha_internal/transaction_batch_impl.hpp [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace shared_model {
+  namespace interface {
+
+    class TransactionBatchImpl : public TransactionBatch {
+     public:
+      explicit TransactionBatchImpl(
+          types::SharedTxsCollectionType transactions);
+
+      TransactionBatchImpl(const TransactionBatchImpl &) = default;
```
Comment (lebdron) [inline shared_model/interfaces/iroha_internal/transaction_batch_impl.hpp]: Not sure if shallow copies are required. It is used anywhere?

File: shared_model/interfaces/iroha_internal/transaction_batch_factory_impl.cpp (hunk 13 above)
Comment (Akvinikym) [reply shared_model/interfaces/iroha_internal/transaction_batch_factory_impl.cpp]: Which exactly validation? Most its part was moved into `transaction_sequence_factory`, though a couple of cases are still left here

File: irohad/ordering/impl/ordering_service_transport_grpc.cpp [hunk 22]
```diff
@@ -38,13 +36,10 @@ grpc::Status OrderingServiceTransportGrpc::onBatch(
           return std::make_shared<shared_model::proto::Transaction>(tx);
         });
 
-    auto batch_result = shared_model::interface::TransactionBatchFactory::
-        createTransactionBatch(
-            txs,
-            shared_model::validation::DefaultSignedTransactionsValidator());
+    auto batch_result = batch_factory_->createTransactionBatch(txs);
```
Comment (l4l) [inline irohad/ordering/impl/ordering_service_transport_grpc.cpp]: The transactions aren't checked for stateless validity here but it probably should since it grpc endpoint

File: shared_model/interfaces/iroha_internal/transaction_batch_factory_impl.cpp [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+  namespace interface {
+    TransactionBatchFactoryImpl::FactoryImplResult
+    TransactionBatchFactoryImpl::createTransactionBatch(
+        const types::SharedTxsCollectionType &transactions) const {
+      std::string reason_name = "Transaction batch factory: ";
+      validation::ReasonsGroupType batch_reason;
+      batch_reason.first = reason_name;
+
+      if (boost::empty(transactions)) {
+        batch_reason.second.emplace_back("Provided transactions list is empty");
```
Comment (lebdron) [inline shared_model/interfaces/iroha_internal/transaction_batch_factory_impl.cpp]: This check will be provided by the parser, so it should not be necessary here.
